### ISSUE TITLE
Receivers client hits memory/storage limits while sending large files

### ISSUE THREAD

**Issue by aswinshenoy** [aswinshenoy/BayJDO#18]
>Can you tell me how it is not (or wont be) happening on the sender's side please? On both sides, chunks are being held, on the receivers end perhaps, its getting progressively build up, while on the sender's side the whole set of chunks (we made from the file) is being held in a JS var.

There is no memory overflow on the senders side. This can be verified by running a stress test and measuring memory usage via dev tools (you can also use a heap snapshot but this will suffice for now). 
![Screenshot_of_test](https://i.imgur.com/NVCIOew.png)

The sender is using about 100MB of memory (consistently) while the receiver is using more than 7GB ( which will increase as more chunks accumulate).
This is because the File object (special kind of blob) returned from the Open File dialog (using [input type="file" ](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/file) ) is not in memory but rather stored in disk. 
You use Dropzone.js which uses the following in its underlying implementation:
 ```html
 <input type="file">
``` 
Here is a snippet from your chunking mechanism:
```js
  _getNextSlice = async () => {
        const slice = this._file.slice(
            this._offset,
            this._offset + this._chunkSize
        );
        return await slice.arrayBuffer().then((buffer) => buffer);
    };
```
The chunk is brought into memory only when `slice.arrayBuffer()` is called. Once the chunk is transmitted to the receiving peer, no further references to it exist. This causes the GC to release the memory previously occupied by the chunk.

_Originally posted by @MidhunSureshR in https://github.com/aswinshenoy/BayJDO/issues/14#issuecomment-668485892_

**Comment by MidhunSureshR** [comment]
I tried to use StreamSaver.js to write chunks into disk as they arrive and you can see the progress [here](https://github.com/MidhunSureshR/BayJDO/tree/18-fix-memory-usage-receiver). It does work but the problem now is that since we don't really store the chunks in memory, we can't really create a video/audio/image preview like before.

**Comment by aswinshenoy** [comment]
> I tried to use StreamSaver.js to write chunks into disk as they arrive and you can see the progress [here](https://github.com/MidhunSureshR/BayJDO/tree/18-fix-memory-usage-receiver). It does work but the problem now is that since we don't really store the chunks in memory, we can't really create a video/audio/image preview like before.

There should be workarounds, easiest one that comes in my mind rn -> The files most likely to be previewed (mp3, images, small videos) are usually going to be small files (<50MB), and in that case we could use the same method as now since there wont be a memory issue etc. When the sender says the file size >50mb (at the beginning), we decide to use the streamsaver technique, otherwise use the same method as now. 

> If you create a new branch I'll open a pull request.
You can open it to merge with master itself, since no multiple projects/contributors are working on this rn, I dont think its necessary to create branches. Please open pull request with headed to the master branch itself

**Comment by MidhunSureshR** [comment]
> When the sender says the file size >50mb (at the beginning), we decide to use the streamsaver technique, otherwise use the same method as now.
Good idea. The only thing left to do is update the react components to not show the preview options if the files are greater than 50MB. I'll open a pull request once it's done.

**Comment by aswinshenoy** [comment]
@MidhunSureshR great! the preview part is only a minor change, you can still open the PR, and I could review, and we could get this into production asap :)
